src/server/game/Entities/Unit/Unit.cpp

```diff
@@ -15684,21 +15684,13 @@ void Unit::ProcDamageAndSpellFor(bool isVictim, Unit* target, uint32 procFlag, u
 
         // Update skills here for players
         // only when you are not fighting other players or their pets/totems (pvp)
-        if (GetTypeId() == TYPEID_PLAYER &&
-                target->GetTypeId() != TYPEID_PLAYER &&
-                !(target->IsTotem() && target->ToTotem()->GetOwner()->IsPlayer()) &&
-                !target->IsPet()
-           )
-        {
-            // On melee based hit/miss/resist/parry/dodge need to update skill (for victim and attacker)
-            if (procExtra & (PROC_EX_NORMAL_HIT | PROC_EX_MISS | PROC_EX_RESIST | PROC_EX_PARRY | PROC_EX_DODGE))
+        if (IsPlayer() && !target->IsCharmedOwnedByPlayerOrPlayer())
+        {
+            // On melee based hit/miss/resist/parry/dodge/block need to update skill (for victim and attacker)
+            if (procExtra & (PROC_EX_NORMAL_HIT | PROC_EX_MISS | PROC_EX_RESIST | PROC_EX_PARRY | PROC_EX_DODGE | PROC_EX_BLOCK))
             {
-                if (target->GetTypeId() != TYPEID_PLAYER)
-                    ToPlayer()->UpdateCombatSkills(target, attType, isVictim, procSpell ? procSpell->m_weaponItem : nullptr);
+                ToPlayer()->UpdateCombatSkills(target, attType, isVictim, procSpell ? procSpell->m_weaponItem : nullptr);
             }
-            // Update defence if player is victim and we block
-            else if (isVictim && procExtra & (PROC_EX_BLOCK))
-                ToPlayer()->UpdateCombatSkills(target, attType, true);
```

Comment:
Why removed? Victims should not skill up their defense skill on block?

Reply:
What makes this `else if` statement different from the `if` above is that it checks for blocked attacks `PROC_EX_BLOCK` and that it additionally checks for `isVictim == true`.

With the condition `isVictim == true`, it will only allow a defence skill up for the player who blocked the attack. The case that player was attacker (`isVictim == false`) is not covered with weaponskill, so currently no chance so skill up weapon when attack was blocked.

To fix that, removing  `isVictim == true` and replacing the hard coded `true` for bool defence with `ToPlayer()->UpdateCombatSkills(target, attType, isVictim );` would cover both cases.

Or just remove the whole block and add `PROC_EX_BLOCK` to the if condition above.

Reply:
`ToPlayer()->UpdateCombatSkills(target, attType, isVictim, procSpell ? procSpell->m_weaponItem : nullptr);` Because now the `isDefense` parameter is always true if it is a victim, But should be true if it is victim AND it's PROC_EX_BLOCK .

Reply:
The if condition above also handles both weapon and defence skill up for `PROC_EX_PARRY | PROC_EX_DODGE`. 
isVictim == true and player did parry/dodge -> defence skill
isVictim == false -> and player's attacks was parried/dodged -> weapon skill

_Each attack which is not evaded has a chance to increase weapon skill, with low weapon skill, ..._

So I think bock should be handled the same way as parry/dodge?

Reply:
Alright. Looks okay